### PR TITLE
feat: add node listing to Kubernetes section

### LINKED ISSUE #7339: Add "Nodes" section for Kubernetes page.
### Is your feature request related to a problem? Please describe

Part of: https://github.com/containers/podman-desktop/issues/7190 / smaller non-epic issue.

Was brought up in a discussion earlier this month of "How do I view my nodes?"

Creating this issue to keep track of adding Nodes to kubernetes dashboard.

### Describe the solution you'd like

Added Nodes section to Kubernetes dashboard similar to kubectl:

```
▶ k get nodes -o wide
NAME   STATUS   ROLES                  AGE     VERSION        INTERNAL-IP    EXTERNAL-IP   OS-IMAGE          KERNEL-VERSION          CONTAINER-RUNTIME
k8s1   Ready    control-plane,master   5d19h   v1.29.4+k3s1   192.168.1.11   <none>        CentOS Stream 9   5.14.0-437.el9.x86_64   containerd://1.7.15-k3s1
k8s2   Ready    <none>                 5d19h   v1.29.4+k3s1   192.168.1.12   <none>        CentOS Stream 9   5.14.0-437.el9.x86_64   containerd://1.7.15-k3s1
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by cdrage**
feat: add node listing to Kubernetes section

### What does this PR do?

* Adds Nodes section to Kubernetes section
* No actions (dont want to delete your node by mistake!). If you want to
  taint / change a role of a node, you can always edit the YAML under
  the Kube tab similar to all the other features
* Labels for control plane / node as per the newest Kubernetes
  guidelines.

### Screenshot / video of UI

Screenshot 2024-05-28 at 12 00 10 PM

https://github.com/containers/podman-desktop/assets/6422176/1adc8df1-f08f-45f6-bb5b-13572577f038

### What issues does this PR fix or reference?

Closes https://github.com/containers/podman-desktop/issues/7339

### How to test this PR?

- [X] Tests are covering the bug fix or the new feature

View the "Nodes" section of the Kubernetes dashboard

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

### CODE REVIEW

**Comment by cdrage** [comment]
IMPORTANT NOTE:

We will need a new icon for the nodes section before merging. Right now it is using the "DeploymentIcon".

@deboer-tim  @ekidneyrh 

From a quick google, Kubernetes uses the "three gears together" to signify a node / control plane.

image

**Comment by feloy** [comment]
Nodes are more frequently not authorized to developers compared to Deployments and similar resources. I think that, if possible, we should check if the informer fails with an Unauthorized error, to avoid retrying the connection infinitely in this case. Or, if the type of error is not known from the informer, to first try with a simple List operation.

This can be part of a follow-up PR

**Comment by cdrage** [comment]
> Nodes are more frequently not authorized to developers compared to Deployments and similar resources. I think that, if possible, we should check if the informer fails with an Unauthorized error, to avoid retrying the connection infinitely in this case. Or, if the type of error is not known from the informer, to first try with a simple List operation.
> 
> 
> 
> This can be part of a follow-up PR

Yup! I agree as well with regards to any form of actions on there. 

It is however nice to see if your node is accessible / online or not (ex.. kind cluster). 

We should indeed add any Unauthorized. 

By default though, aren't most clusters the NodeList is readable?

**Comment by feloy** [comment]
> > Nodes are more frequently not authorized to developers compared to Deployments and similar resources. I think that, if possible, we should check if the informer fails with an Unauthorized error, to avoid retrying the connection infinitely in this case. Or, if the type of error is not known from the informer, to first try with a simple List operation.
> > This can be part of a follow-up PR
> 
> Yup! I agree as well with regards to any form of actions on there.
> 
> It is however nice to see if your node is accessible / online or not (ex.. kind cluster).
> 
> We should indeed add any Unauthorized.
> 
> By default though, aren't most clusters the NodeList is readable?

It is not the case for example for Openshift Sandbox:

```
$ kubectl get pods       
No resources found in phmartin-dev namespace.
$ kubectl get nodes
Error from server (Forbidden): nodes is forbidden: User "phmartin" cannot list resource "nodes" in API group "" at the cluster scope
```

**Comment by cdrage** [comment]
> > > Nodes are more frequently not authorized to developers compared to Deployments and similar resources. I think that, if possible, we should check if the informer fails with an Unauthorized error, to avoid retrying the connection infinitely in this case. Or, if the type of error is not known from the informer, to first try with a simple List operation.
> > > This can be part of a follow-up PR
> > 
> > 
> > Yup! I agree as well with regards to any form of actions on there.
> > It is however nice to see if your node is accessible / online or not (ex.. kind cluster).
> > We should indeed add any Unauthorized.
> > By default though, aren't most clusters the NodeList is readable?
> 
> It is not the case for example for Openshift Sandbox:
> 
> ```
> $ kubectl get pods       
> No resources found in phmartin-dev namespace.
> $ kubectl get nodes
> Error from server (Forbidden): nodes is forbidden: User "phmartin" cannot list resource "nodes" in API group "" at the cluster scope
> ```

So this was pretty straightforward to implement!

I created a check that if we are connected to a cluster, but no nodes are listed (tested via red hat sandbox), we will output a different error that the user may not have permission to view them, which would be normal.

I have added a test as well as updated the page:
Screenshot 2024-05-29 at 11 13 59 AM

**Comment by ekidneyrh** [comment]
Hi @cdrage, will I open a ticket for the icon? Also, is the icon for the Nav bar / empty screen page?

**Comment by cdrage** [comment]
> Hi @cdrage, will I open a ticket for the icon? Also, is the icon for the Nav bar / empty screen page? 

Yes please! This issue was unplanned / not in the sprint. So if you have the bandwidth to create the icon that'd be awesome.

**Comment by ekidneyrh** [comment]
https://github.com/containers/podman-desktop/issues/7378 logged here :) also wondering will this be displayed in a table also? Eg will we need a filled version too.

**Comment by cdrage** [comment]
> #7378 logged here :) also wondering will this be displayed in a table also? Eg will we need a filled version too.

Yup! It'll be displayed in the table too. I'll have to make sure to remind myself to update that

File: packages/main/src/plugin/kubernetes-context-state.ts [hunk 1]
```diff
@@ -361,7 +365,8 @@ export class ContextsManager {
   }
 
   isContextInKubeconfig(context: KubeContext): boolean {
-    // if there is no cluster on the kubeconfig with the value of context.cluster -> false
+    // if there is no cluster on the kubeconfig with:w
```
Comment (feloy) [inline packages/main/src/plugin/kubernetes-context-state.ts]: vim typo? ;)

Comment (cdrage) [reply packages/main/src/plugin/kubernetes-context-state.ts]: Yup haha, I use the plugin in VSCode

**Comment by cdrage** [comment]
To improve the UI, I also added some mini-icons for representing control plane, node:

Screenshot 2024-05-30 at 2 52 01 PM

**Comment by cdrage** [comment]
Screenshot 2024-05-30 at 3 10 11 PM

Minor update, added an icon to represent control plane / node presentation better.

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+    "vitest.disableWorkspaceWarning": true
+}
```
Comment (feloy) [inline .vscode/settings.json]: Is it added to the repo on purpose?

**Review by feloy — COMMENTED** [review]
LGTM, except some minor tweaks

The node icon seems too much on the left (not horizontally centered), and seems truncated on the left side
node_icon
no_node

**Comment by cdrage** [comment]
> LGTM, except some minor tweaks
> 
> The node icon seems too much on the left (not horizontally centered), and seems truncated on the left side node_icon no_node

Thank you! That's on purpose as it's using the "settings icon". Once we have a new nodes icon (@ekidneyrh) I can update it.

**Review by feloy — APPROVED** [review]

**Review by axel7083 — COMMENTED** [review]
image

Amazing job, just one question: Could we have the `text ellipsis` for the name ?

**Comment by cdrage** [comment]
> image
> 
> Amazing job, just one question: Could we have the `text ellipsis` for the name ?

Opened up https://github.com/containers/podman-desktop/issues/7498 for this as it appears in multiple sections / will close in a separate PR.

**Comment by cdrage** [comment]
Since this is a large PR and the only thing that is missing is the updated icon, I'm going to set this to auto-merge, also to avoid any large rebasing (as this is 1,200 lines).

Follow up issue is: https://github.com/containers/podman-desktop/issues/7378 which should go in before release.

**Pull request merged**